app/dashboard/page.tsx

```diff
@@ -1,135 +1,220 @@
   const [userId, setUserId] = useState<string | null>(null);
-  const [devices, setDevices] = useState<any[]>([]);
+  const [devices, setDevices] = useState<Device[]>([]);
   const [selectedDeviceId, setSelectedDeviceId] = useState<string>("");
-  const [selectedDevice, setSelectedDevice] = useState<any | null>(null);
-  const [editValues, setEditValues] = useState({
+  const [selectedDevice, setSelectedDevice] = useState<Device | null>(null);
+  const [editValues, setEditValues] = useState<Omit<Device, "_id">>({
     name: "",
     color: "Red",
     brightness: 100,
     powered: false,
   });
   const [loading, setLoading] = useState(true);
 
+  const [colorHex, setColorHex] = useState("#ff0000");
+  const [colorRgb, setColorRgb] = useState<RGB>({ r: 255, g: 0, b: 0 });
+  const [hexInput, setHexInput] = useState("#ff0000");
+  const lastChangeSource = useRef<"hex" | "rgb" | null>(null);
+  const hiddenColorInput = useRef<HTMLInputElement | null>(null);
+
+  // ------------------ Fetch User + Devices ------------------
   useEffect(() => {
     const fetchUserData = async () => {
       try {
-        const response = await axios.get("/api/users/me");
+        const response = await axios.get<{ data: User }>("/api/users/me");
         const user = response.data.data;
 
         if (user && user._id) {
           setUserId(user._id);
-          const deviceRes = await axios.get("/api/users/dashboard");
+          const deviceRes = await axios.get<{ devices: Device[] }>("/api/users/dashboard");
           setDevices(deviceRes.data.devices);
         } else {
           throw new Error("User ID not found");
         }
-      } catch (error: any) {
-        console.error("Error fetching data:", error.message);
+      } catch (error: unknown) {
+        if (error instanceof Error) console.error("Error fetching data:", error.message);
         router.push("/login");
       } finally {
         setLoading(false);
       }
     };
-
     fetchUserData();
   }, [router]);
 
+  // ------------------ Selected Device ------------------
   useEffect(() => {
-    if (selectedDeviceId) {
-      const device = devices.find((d) => d._id === selectedDeviceId);
-      setSelectedDevice(device);
+    if (!selectedDeviceId) return;
+
+    const device = devices.find(d => d._id === selectedDeviceId) || null;
+    setSelectedDevice(device);
+
+    if (device) {
       setEditValues({
-        name: device?.name || "",
-        color: device?.color || "Red",
-        brightness: device?.brightness || 100,
-        powered: device?.powered || false,
+        name: device.name,
+        color: device.color,
+        brightness: device.brightness,
+        powered: device.powered,
       });
+
+      const hex = colorMap[device.color] || "#ff0000";
+      setColorHex(hex);
+      setColorRgb(hexToRgb(hex)!);
+      setHexInput(hex);
     }
   }, [selectedDeviceId, devices]);
 
+  // ------------------ HEX ↔ RGB Sync ------------------
+  useEffect(() => {
```

Comment:
`useEffect` will rerender the page since you are calling `setHexInput`.  This will cause the app to bottleneck.  

1. You can either memoize the `colorHex` data
2. Or the easiest option is to call `setHexInput` with the `colorHex` data once the user lets goes of the mouse.  Look at https://www.w3schools.com/jsref/event_onmouseup.asp for more details.